File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/domain/common/configuration/GetConfigurationProcess.java
package com.crpr.androidcinema.domain.common.configuration;

import com.crpr.androidcinema.data.api.models.ApiConfiguration;
import com.crpr.androidcinema.domain.common.Result;
import com.crpr.androidcinema.domain.common.providers.ImageUrlProvider;

import rx.Observable;

/**
 * Created by claudioribeiro on 09/07/16.
 */
public class GetConfigurationProcess implements GetConfiguration.Process {

    private final GetConfiguration.Service _service;

    public GetConfigurationProcess(GetConfiguration.Service service){
        this._service = service;
    }

    /**************** GET UPDATED CONFIGURATION ***********************/
    public Observable<Result> getConfiguration(){
        return _service.getConfiguration()
                    .flatMap(this::configureImageUrlProvider);
    }

    private Observable<Result> configureImageUrlProvider(ApiConfiguration configuration) {

        ConfigurationModel model = ConfigurationModel
                                    .url(configuration.getImagesConfiguration().getBaseUrl())
                                    .secureUrl(configuration.getImagesConfiguration().getSecureBaseUrl())
                                    .build();

        if(model == null){
            return Observable.just(new Result(Result.ERROR));
        }

        ImageUrlProvider.sharedInstance().setCurrentConfig(model);

        return Observable.just(new Result(Result.OK));
    }
}

File: AndroidCinema/app/src/main/java/com/crpr/androidcinema/domain/common/configuration/GetConfigurationProcess.kt
package com.crpr.androidcinema.domain.common.configuration

import com.crpr.androidcinema.data.api.models.ApiConfiguration
import com.crpr.androidcinema.domain.common.Result
import com.crpr.androidcinema.domain.common.providers.ImageUrlProvider
import rx.Observable

/**
 * Created by claudioribeiro on 09/07/16.
 */
class GetConfigurationProcess(private val _service: GetConfiguration.Service) : GetConfiguration.Process {

    /**************** GET UPDATED CONFIGURATION  */
    override val configuration: Observable<Result>
        get() = _service.configuration
                .flatMap<Result>({ this.configureImageUrlProvider(it) })

    private fun configureImageUrlProvider(configuration: ApiConfiguration): Observable<Result> {

        val model = ConfigurationModel
                .url(configuration.images.base_url!!)
                .secureUrl(configuration.images.secure_base_url!!)
                .build()

        ImageUrlProvider.sharedInstance().setCurrentConfig(model)

        return Observable.just(Result(Result.OK))
    }
}
